src/shim/preload_libraries.c

```diff
@@ -419,8 +450,14 @@ int getaddrinfo(const char* node, const char* service,
         // TODO: look for IPv6 addresses in /etc/hosts.
     }
     if (add_ipv4) {
-        _getaddrinfo_add_matching_hosts_ipv4(
-            res, &tail, node, add_tcp, add_udp, add_raw, port);
+        // Try a Shadow syscall first to avoid scanning the /etc/hosts file.
+        uint32_t addr;
+        if (_syscall_hostname_to_addr_ipv4(node, &addr)) {
+            _getaddrinfo_appendv4(res, &tail, add_tcp, add_udp, add_raw, addr, port);
+        } else {
+            // Fall back to scanning /etc/hosts.
+            _getaddrinfo_add_matching_hosts_ipv4(res, &tail, node, add_tcp, add_udp, add_raw, port);
```

Comment:
Do we expect to have to fall back to the hosts file in any "normal" case, or is this an attempt to recover from a bug in shadow? Should there be a warning log entry here?

Reply:
I've never seen the new syscall fail, I just didn't want to remove the /etc/hosts scanning functionality yet. I'll add a warning, I think that's appropriate.